libmpv/stream_cb.h

```diff
@@ -0,0 +1,164 @@
+ * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ */
+
+/*
+ * Note: the client API is licensed under ISC (see above) to ease
+ * interoperability with other licenses. But keep in mind that the
+ * mpv core is still mostly GPLv2+. It's up to lawyers to decide
+ * whether applications using this API are affected by the GPL.
+ * One argument against this is that proprietary applications
+ * using mplayer in slave mode is apparently tolerated, and this
+ * API is basically equivalent to slave mode.
+ */
+
+#ifndef MPV_CLIENT_API_STREAM_CB_H_
+#define MPV_CLIENT_API_STREAM_CB_H_
+
+#include "client.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * Warning: this API is not stable yet.
+ *
+ * Overview
+ * --------
+ *
+ * This API can be used to make mpv read from a stream with a custom
+ * implementation. This interface is inspired by funopen on BSD and
+ * fopencookie on linux. The stream is backed by user-defined callbacks
+ * which can implement customized open, read, seek, size and close behaviors.
+ *
+ * Usage
+ * -----
+ *
+ * Create the sub-api with mpv_get_sub_api(), then initialize it with
+ * mpv_stream_cb_init(). Register your stream callbacks with the
+ * mpv_stream_cb_set_*_fn(). Note that your custom callbacks must not
+ * invoke libmpv APIs as that would cause a deadlock.
+ *
+ * Once registered, you can `loadfile cb://myfile`. Your open_fn will be invoked
+ * with the URI and must return an opaque cookie. This cookie will be passed as
+ * the first argument to all the remaining stream callbacks.
+ */
+
+/**
+ * Opaque context, returned by mpv_get_sub_api(MPV_SUB_API_STREAM_CB).
+ *
+ * This context is global to libmpv, and the callbacks set within it are used
+ * by all mpv_handles created in the same process.
+ */
+typedef struct mpv_stream_cb_context mpv_stream_cb_context;
+
+/**
+ * Open callback used to implement a custom stream.
+ *
+ * @param user_data opaque user data provided via mpv_stream_cb_init()
+ * @param uri name of the stream to be opened
+ * @return opaque cookie identifing the newly opened stream
+ * @return NULL if the URI cannot be opened.
+ */
+typedef void *(*mpv_stream_cb_open_fn)(void *user_data, char *uri);
+
+/**
+ * Read callback used to implement a custom stream.
+ * The semantics of the callback match read(2).
+ *
+ * @param cookie opaque cookie identifying the stream,
+ *               returned from mpv_stream_cb_open_fn
+ * @param buf buffer to read data into
+ * @param size of the buffer
+ * @return number of bytes read into the buffer
+ * @return 0 on EOF
+ * @return -1 on error
+ */
+typedef int64_t (*mpv_stream_cb_read_fn)(void *cookie, char *buf, uint64_t nbytes);
+
+/**
+ * Seek callback used to implement a custom stream.
+ * The semantics of this callback match lseek(2).
```

Comment:
This is different from POSIX for certain. The function can return an error if the seek position is beyond the end of the file.